src/granite_orm/validators.cr

```diff
@@ -1,34 +1,48 @@
 require "./error"
 
+# Analyze validation blocks and procs
+#
+# By example:
+# ```
+# validate :name, "can't be blank" do |user|
+#   !user.name.to_s.blank?
+# end
+#
+# validate :name, "can't be blank", -> (user : User) do
+#   !user.name.to_s.blank?
+# end
+#
+# name_required = ->(model : Granite::ORM::Base) { !model.name.to_s.blank? }
+# validate :name, "can't be blank", name_required
+# ```
 module Granite::ORM::Validators
   getter errors = [] of Error
 
   macro included
     macro inherited
-      @validators = Array({field: Symbol, message: String, block: Proc(Bool)}).new
-      def validate!
+      @@validators = Array({field: String, message: String, block: Proc(self, Bool)}).new
+
+      def self.validate(message : String, &block : self -> Bool)
+        self.validate(:base, message, block)
       end
-    end
-  end
 
-  macro validate(message)
-    def validate!
-      previous_def
-      @validators << {field: :base, message: {{message}}, block: ->{{{yield}}}}
-    end
-  end
+      def self.validate(field : (Symbol | String), message : String, &block : self -> Bool)
+        self.validate(field, message, block)
+      end
+
+      def self.validate(message : String, block : self -> Bool)
+        self.validate(:base, message, block)
+      end
 
-  macro validate(field, message)
-    def validate!
-      previous_def
-      @validators << {field: {{field}}, message: {{message}}, block: ->{{{yield}}}}
+      def self.validate(field : (Symbol | String), message : String, block : self -> Bool)
+        @@validators << {field: field.to_s, message: message, block: block}
```

Comment:
doesn't this recreate the "Error on one instance, error on all instances" bug?

Reply:
I tested it and surprisingly works without that issue 😅